python/cudf_polars/cudf_polars/experimental/groupby.py

```diff
@@ -81,14 +84,16 @@ def decompose(
         The initial aggregation expressions.
     list[NamedExpr]
         The reduction expressions.
+    bool
+        Whether we need to pre-shuffle on the group_by keys.
```

Comment:
How would one know what types of operations need a pre-shuffle?

Reply:
I'm not sure how many operations require a pre-shuffle. In this case, the partition-wise `n_unique` calculation does not preserve the original unique values, so we don't know if we are double-counting unique values during a tree reduction unless we were already shuffled by the group_by keys **before** the original partition-wise calculation.